elf2flt: add riscv 64-bits support

damien-lemoal opened the pull request:
Add support for riscv 64bits ISA by defining the relocation types
R_RISCV_32_PCREL, R_RISCV_ADD32, R_RISCV_SUB32, R_RISCV_32 and
R_RISCV_64. riscv64 support also needs the __global_pointer$ symbol to
be defined right after the relocation tables in the data section.
Furthermore, the .got and .got.plt sections must be reversed. These 2
requirements are handled with runtime modifications of the default
linker script using the append_sed() function.
(1) For the .got.plt and .got sections order swap, append_sed() is used
to rename "(.got.plt)" to "(.got.tmp)" and to rename "(.got)" to
"(.got.plt)". A last call finalize the name swap by replacing
"(.got.tmp)" with "(.got)"
(2) For the global pointer synbol, a definition line starting with
"RV64_GP" is added. The "RV64_GP" string is removed if the target CPU
type is riscv64. The definition line is dropped for other CPU types.

With these changes, buildroot/busybox builds and run on NOMMU
systems with kernel 5.13. Tested on Canaan Kendryte K210 boards.

This patch is based on earlier work by Christoph Hellwig <hch@lst.de>.

Signed-off-by: Damien Le Moal <damien.lemoal@wdc.com>

damien-lemoal committed:
elf2flt: add riscv 64-bits support

Add support for riscv 64bits ISA by defining the relocation types
R_RISCV_32_PCREL, R_RISCV_ADD32, R_RISCV_SUB32, R_RISCV_32 and
R_RISCV_64. riscv64 support also needs the __global_pointer$ symbol to
be defined right after the relocation tables in the data section. To
define this symbol, the "RISCV_GP" line prefix is added. The "RISCV_GP"
string is removed if the target CPU type is riscv64 and the definition
line is dropped for other CPU types.

With these changes, buildroot and busybox build and run on riscv NOMMU
systems with Linux kernel including patch 6045ab5fea4c
("binfmt_flat: do not stop relocating GOT entries prematurely on riscv")
fixing the binfmt_flat loader. Tested on QEMU and Canaan Kendryte K210
boards.

This patch is based on earlier work by Christoph Hellwig <hch@lst.de>.

Signed-off-by: Damien Le Moal <damien.lemoal@wdc.com>

damien-lemoal commented:
Notes:
* This is an RFC as I am still not entirely sure if this is acceptable. E.g., I cannot get rid of the section ordering problem. ".got" and ".got.plt" need to be reversed. I do not understand why.
* This still does not include support for RISCV 32-bits as I have not yet done any tests with a 32-bits NOMMU kernel (unsure if it even works).

Comments are more than welcome !

vapier commented on elf2flt.ld.in:
```diff
@@ -109,6 +109,7 @@ W_RODAT:	*(.gnu.linkonce.r*)
 		. = ALIGN(0x20) ;
 		LONG(-1)
 		. = ALIGN(0x20) ;
+RV64_GP		__global_pointer$ = . + 0x800;
```
use a `:` after it to match the style of all the others

vapier commented on ld-elf2flt.c:
```diff
@@ -324,6 +324,24 @@ static int do_final_link(void)
+		append_sed(&sed, "(.got.tmp)", "(.got.plt)");
+
+		/*
+		 * The global pointer symbol is defined after the GOT.
```
one-line this comment block

vapier commented on ld-elf2flt.c:
```diff
@@ -324,6 +324,24 @@ static int do_final_link(void)
+		 */
+		append_sed(&sed, "^RV64_GP", "");
+	} else {
+		/* Get rid of the global pointer definition */
```
period at the end of sentences

vapier commented on ld-elf2flt.c:
```diff
@@ -324,6 +324,24 @@ static int do_final_link(void)
+		append_sed(&sed, "^RV64_GP", "");
+	} else {
+		/* Get rid of the global pointer definition */
+		append_sed(&sed, "^RV64_GP", NULL);
```
iiuc, `__global_pointer$` is not specific to riscv64, but all riscv ABIs utilize it.  so i would name it `RISCV_GP`.

vapier reviewed (commented):
is there an open bug either with the riscv or binutils folks to figure out the got/plt ordering ?  it's great that we've managed to reduce this down to just the ordering of those two sections.

damien-lemoal commented:
Thanks for the comment. I will update the patch. I am still doing a little more digging to see if I can remove this section reordering as I really do not see why it is necessary.

damien-lemoal commented:
PR updated. I addressed all comments and also fixed a bug with some relocation types handling. With this patch both buildroot and manual busybox NOMMU builds for RISC-V 64-bits work using 5.13+kernels.
Reviews are most welcome !

vapier commented:
can you go through and mark conversations as "resolved" ?

and can you cover the question wrt open bug report somewhere for us to link to ?

damien-lemoal commented:
Previous comments are all marked as "resolved" now. I am not sure what you mean with "open bug report" ?
This support is not yet upstream anywhere. My primary target is buildroot and the patch series I submitted that uses this elf2flt patch is still pending (I just posted v5 today with this new version of the elf2flt patch). The "bug" with the relocation types I mentioned in my previous comment today is something I found while working on the buildroot patches: sometimes, the buildroot image would not fully boot, with busybox exiting immediately on startup. The problem was that PC relative relocations handled by the loader where not skipped here ("break" statement instead of "continue" in the switch-case within the relocation loop. Fixing that fixed the buildroot issue I was seeing. With that fixed and tested, I updated this elf2flt patch and v5 of the buildroot series (adding support for NO-MMU builds for various boards using the Canaan K210 SoC).

Does this answer your question ?

vapier commented on ld-elf2flt.c:
```diff
@@ -324,6 +324,22 @@ static int do_final_link(void)
+	if (streq(TARGET_CPU, "riscv64")) {
+		/*
+		 * The .got section must come before the .got.plt section
+		 * (gcc/ld bug ?).
```
where is the bug/issue tracker for this ?

damien-lemoal replied:
I am not aware of anything about that. And that seems to be an issue only with flatbin since I have not seen any problem with regular MMU elf code. I will try to dig further and understand what is going on.

vapier commented on elf2flt.c:
```diff
@@ -1832,6 +1846,15 @@ int main(int argc, char *argv[])
   if (!load_to_ram && !pfile)
     load_to_ram = 1;
 
+#if defined(TARGET_riscv64)
```
i mentioned in a previous review somewhere that we should just omit this as basically unncessary.  the default is to use `-r` already, and no one runs this tool directly by hand

damien-lemoal replied:
Oops. Yes indeed. Buildroot was fixed to add the -r option but I forgot to remove this. Will respin the patch without this.

Commit:
elf2flt.ld: reinstate 32 byte alignment for .data section

Commit 8a3e74446fe7 ("allow to build arm flat binaries") moved the
following commands:
	. = ALIGN(0x20) ;
	@SYMBOL_PREFIX@_etext = . ;
from the .text section to the top level in the SECTIONS node.

The .text output section is being directed to a memory region using the
"> flatmem :text" output section attribute. Commands in the top level in
the SECTIONS node are not.

This means that the ALIGN() command is no longer being appended to the
flatmem memory region, it will simply update the Location Counter.

The heuristic for placing an output section is described here:
https://sourceware.org/binutils/docs-2.38/ld.html#Output-Section-Address

"If an output memory region is set for the section then it is added to this
region and its address will be the next free address in that region."

Since the .data section is being directed to the same memory region as the
.text section, this means that the Location Counter is not used when
assigning an address to the .data output section, it will simply use the
next free address.

No longer directing these commands to the flatmem memory region means that
the .data output section is no longer aligned to a 32 byte boundary.

Before commit 8a3e74446fe7 ("allow to build arm flat binaries"):
$ readelf -S busybox_unstripped.gdb | grep data
  [ 3] .data             PROGBITS         0000000000035ac0  00036ac0
$ readelf -s busybox_unstripped.gdb | grep _etext
 19286: 0000000000035ac0     0 NOTYPE  GLOBAL DEFAULT    1 _etext

After commit 8a3e74446fe7 ("allow to build arm flat binaries"):
$ readelf -S busybox_unstripped.gdb | grep data
  [ 3] .data             PROGBITS         0000000000035ab0  00036ab0
$ readelf -s busybox_unstripped.gdb | grep _etext
 19287: 0000000000035ac0     0 NOTYPE  GLOBAL DEFAULT    3 _etext

The .data output section has to be aligned to a 32 byte boundary, see the
FLAT_DATA_ALIGN 0x20 macro and its usage in fs/binfmt_flat.c:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/binfmt_flat.c?h=v5.17#n59

Readd an explicit ALIGN attribute on the .data section itself, since the
linker will obey this attribute regardless if being directed to a memory
region or not. Also remove the ALIGN() command before the .data section,
since this misleads the reader to think that the Location Counter is used
when assigning an address to the .data section, when it actually is not.

Fixes: 8a3e74446fe7 ("allow to build arm flat binaries")
Signed-off-by: Niklas Cassel <niklas.cassel@wdc.com>

damien-lemoal commented:
PR update: This update adds one patch from Niklas to fix alignment of the data section in the elf2flt.ld.in file. The 32 B alignment is required by Linux kernel binfmt_flat loader. The second patch is also updated to remove reversing of the got sections. The root cause of the problem that this was (badly) fixing was in fact in the kernel binfmt_flat loader. A patch on the kernel side solves the issue and reversing the got and got.plt sections is no longer necessary.

With these patches added to buildroot, NO-MMU riscv64 support is now stable. Tested with QEMU and on Canaan Kendryte K210 boards.

Please consider these patches for addition.

Commit:
elf2flt: fix .eh_frame section handling

elf2flt.ld linker script positions the .eh_frame section in an output
section after the .text and .data output sections.

However, when elf2flt.c is supplied the ELF linked using the elf2flt.ld
linker script, it only looks at the flags for each input section, and
then puts it in either a bFLT text, data or bss output section.

Commit ba379d08bb7 ("elf2flt: fix for segfault on some ARM ELFs")
modified the section scanning loop of elf2flt main() function to put
read-only relocation data sections in the bFLT text output section so
that the .ARM.exidx section is placed in the .text flat output section.
Previously a read-only relocation data section would be put in the data
output section.

On ARM, the .eh_frame section does not have the SEC_RELOC flag set, so
it will still end up in the data output section. However, on
architectures that generates the .eh_frame section with the SEC_RELOC
flag set, this section will now be placed in the text output section.

The logic in elf2flt will handle all sections in order, and since the
input order is .text, .data, and .eh_frame, putting .eh_frame in text
output section does not work, since elf2flt.c has already put the .data
input section in the bFLT data output section. This leads to the
following print (example for riscv64 architecture):

buildroot/output/host/riscv64-buildroot-linux-uclibc/bin/elf2flt:
ERROR: text=0x3bab8 overlaps data=0x33f60 ?

The way that elf2flt is written, we cannot append to the text output
section after an input section has been added to the data output
section. It might be possible to change this, but that would require
moving all the code the was already placed in the data output section.

Instead, let's allow putting a read-only relocation data section in the
text output section (so that .ARM.exidx will still be placed correctly),
but only if there has not yet been anything placed in the data output
section.

That way .ARM.exidx will still be placed correctly, and .eh_frame will
be placed correctly in the .data output section, regardless if it has
flag SEC_RELOC set or not.

Fixes: ba379d08bb7 ("elf2flt: fix for segfault on some ARM ELFs")
Signed-off-by: Damien Le Moal <damien.lemoal@opensource.wdc.com>
Signed-off-by: Niklas Cassel <niklas.cassel@wdc.com>

damien-lemoal commented:
PR updated to add one more patch to better handle .eh_frame section. Feed back welcome !

gregungerer commented:
RISC-V support has been merged into main in later versions of this PR.
